hubtools/src/archive_builder.rs

```diff
@@ -0,0 +1,162 @@
+        }
+    }
+
+    /// Construct a fake `RawHubrisImage` that is not valid, but _appears_ to be
+    /// valid: it contains the correct magic numbers at the correct offsets to
+    /// look like a hubris image and contain a caboose.
+    pub fn with_fake_image() -> Self {
+        const FAKE_IMAGE_SIZE: usize = 1024;
+        const FAKE_CABOOSE_SIZE: usize = 256;
+
+        // Built a dummy image that still looks reasonable enough for testing
+        // (e.g, we stick some magic values in the right places and allow for
+        // insertion of a caboose).
+        let mut data = io::Cursor::new(vec![0; FAKE_IMAGE_SIZE]);
+
+        // Write the header magic at the largest possible value where
+        // `RawHubrisArchive` might look for it.
+        data.set_position(u64::from(
+            header::POSSIBLE_OFFSETS.iter().copied().max().unwrap(),
+        ));
+        data.write_all(&header::MAGIC.last().copied().unwrap().to_le_bytes())
+            .unwrap();
+
+        // Write the image size immediately after the header magic.
+        data.write_all(&(FAKE_IMAGE_SIZE as u32).to_le_bytes())
+            .unwrap();
+
+        // Write the caboose magic at the beginning of the caboose.
+        data.set_position((FAKE_IMAGE_SIZE - FAKE_CABOOSE_SIZE) as u64);
+        data.write_all(&CABOOSE_MAGIC.to_le_bytes()).unwrap();
+
+        // Write the caboose size as the last word in the image.
+        data.set_position((FAKE_IMAGE_SIZE - 4) as u64);
+        data.write_all(&(FAKE_CABOOSE_SIZE as u32).to_le_bytes())
+            .unwrap();
+
+        let image =
+            RawHubrisImage::from_binary(data.into_inner(), 0, 0).unwrap();
+
+        // Ensure our caboose staging above was correct.
+        assert_eq!(
+            (FAKE_IMAGE_SIZE - FAKE_CABOOSE_SIZE + 4) as u32
+                ..FAKE_IMAGE_SIZE as u32 - 4,
+            image.caboose_range().unwrap()
+        );
+
+        Self::with_image(image)
+    }
+
+    /// Set the version number including in the zip file comment.
+    pub fn archive_version(&mut self, archive_version: u32) -> &mut Self {
+        self.archive_version = archive_version;
+        self
+    }
+
+    /// Overwrite the caboose of the image included in this archive.
+    pub fn write_caboose(&mut self, data: &[u8]) -> Result<&mut Self, Error> {
+        self.image.write_caboose(data)?;
+        Ok(self)
+    }
+
+    pub fn build<W>(self, out: W) -> Result<W, Error>
+    where
+        W: io::Write + io::Seek,
+    {
+        let opts = zip::write::FileOptions::default()
+            .compression_method(zip::CompressionMethod::Bzip2);
+
+        let mut archive = ZipWriter::new(out);
+        archive.set_comment(format!(
+            "hubris build archive v{}",
+            self.archive_version
+        ));
+
+        archive.start_file("img/final.elf", opts)?;
+        let elf = self.image.to_elf()?;
+        archive.write_all(&elf).unwrap();
+        let bin = self.image.to_binary()?;
```

Comment:
Missing a call to `archive.set_file("img/final.bin", opts)?` here?

Reply:
Whoops, yes. Good catch.